### PR TITLE
seraphis checkpoint cache

### PR DESCRIPTION
**Pull request by ghostway0**
This has some changes it has to go through, certainly.

### CODE REVIEW

File: tests/unit_tests/seraphis_cache.cpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+                uint64_t middle = window.at(window.size() / 2);
+                m_checkpoints.erase(middle);
+            }
+
+            m_checkpoints.erase(*window.begin());
+        }
+    }
+}
+
+TEST(seraphis_cache, main)
```
Comment (ghostway0) [inline tests/unit_tests/seraphis_cache.cpp]: this has to be written

File: tests/unit_tests/seraphis_cache.cpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    for (size_t i = 0; i < block_ids.size(); i++)
+    {
+        uint64_t current_height = first_block_height + i;
+        m_checkpoints.insert({current_height, block_ids[i]});
+    }
+
+    clean_prunable_checkpoints();
+}
+
+void Cache::clean_prunable_checkpoints()
```
Comment (ghostway0) [inline tests/unit_tests/seraphis_cache.cpp:45]: I dont like how this looks

File: tests/unit_tests/seraphis_cache.cpp [hunk 3]
```diff
@@ -0,0 +1,74 @@
+#include "seraphis_cache.h"
+
+#include <boost/range/adaptor/reversed.hpp>
+#include <gtest/gtest.h>
+
+bool HalfPruning::should_prune(const std::deque<uint64_t> &window, uint64_t chain_top)
```
Comment (ghostway0) [inline tests/unit_tests/seraphis_cache.cpp]: will not stay as a virtual, would probably be incorporated into the Cache, just wanted to open options

File: tests/unit_tests/seraphis_cache.cpp [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+{
+    if (chain_top - *window.end() <= m_num_unprunable)
```
Comment (UkoeHB) [inline tests/unit_tests/seraphis_cache.cpp]: [dereferencing `.end()` is UB](https://en.cppreference.com/w/cpp/container/deque/end), do you mean `.back()`? Calling `.back()` on an empty container is also [UB](https://en.cppreference.com/w/cpp/container/deque/back), and you don't test the size here.

Comment (UkoeHB) [reply tests/unit_tests/seraphis_cache.cpp]: I think this will cause the unprunable region to be larger than your config specifies, since if the middle of the range is pruned then the region between the middle and the end will never be pruned in the case where this condition is right on the edge.

EDIT: or maybe smaller than the config, since it looks like the window is sorted backward?

File: tests/unit_tests/seraphis_cache.cpp [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        return false;
+
+    uint64_t width = *window.end() - *window.begin();
+    return window.size() * m_max_separation > width;
+}
+
+uint64_t Cache::get_nearest_block_height_clampdown(uint64_t test_height)
```
Comment (UkoeHB) [inline tests/unit_tests/seraphis_cache.cpp]:
```suggestion
uint64_t Cache::get_nearest_block_height_clampdown(const uint64_t test_height)
```

Mark parameters `const` if they are supposed to be immutable.

File: tests/unit_tests/seraphis_cache.cpp [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+{
+    if (test_height < m_refresh_height)
+        return (uint64_t)(-1);
+
+    auto checkpoint = m_checkpoints.upper_bound(test_height);
+    if (checkpoint == m_checkpoints.end())
+        return (uint64_t)(-1);
+
+    return (*--checkpoint).first;
```
Comment (UkoeHB) [inline tests/unit_tests/seraphis_cache.cpp:19]: UB if checkpoint is `.begin()`

File: tests/unit_tests/seraphis_cache.cpp [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+void Cache::insert_new_block_ids(uint64_t first_block_height, std::vector<crypto::hash> block_ids)
```
Comment (UkoeHB) [inline tests/unit_tests/seraphis_cache.cpp]: Why passing `block_ids` by value?

File: tests/unit_tests/seraphis_cache.cpp [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+{
+    auto erase_begin = m_checkpoints.lower_bound(first_block_height);
```
Comment (UkoeHB) [inline tests/unit_tests/seraphis_cache.cpp:29]: I like this trick with `lower_bound()`, did not know about this function before.

File: tests/unit_tests/seraphis_cache.cpp [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    if (erase_begin != m_checkpoints.end())
+    {
+        m_checkpoints.erase(erase_begin, m_checkpoints.end());
+    }
+
+    for (size_t i = 0; i < block_ids.size(); i++)
+    {
+        uint64_t current_height = first_block_height + i;
+        m_checkpoints.insert({current_height, block_ids[i]});
```
Comment (UkoeHB) [inline tests/unit_tests/seraphis_cache.cpp]:
```suggestion
        m_checkpoints[current_height] = block_ids[i];
```
This might be more efficient, not sure (it's at least a bit less verbose).

File: tests/unit_tests/seraphis_cache.cpp [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    clean_prunable_checkpoints();
```
Comment (UkoeHB) [inline tests/unit_tests/seraphis_cache.cpp]:
```suggestion
    this->clean_prunable_checkpoints();
```

I like to add `this->` to internal member function calls to disambiguate them from free function calls.

File: tests/unit_tests/seraphis_cache.cpp [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+void Cache::clean_prunable_checkpoints()
+{
+    uint64_t latest_height = m_checkpoints.end()->first;
```
Comment (UkoeHB) [inline tests/unit_tests/seraphis_cache.cpp]: A) dereferencing `.end()` is UB (can use `.rbegin()` instead)
B) you need to check the size

File: tests/unit_tests/seraphis_cache.cpp [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+    std::deque<uint64_t> window{};
+    window.resize(m_window_size / 2 - 2, m_refresh_height - m_max_separation);
```
Comment (UkoeHB) [inline tests/unit_tests/seraphis_cache.cpp]: What if window size is < 4, or refresh height < max separation? I don't really understand what this initialization is trying to do.

File: tests/unit_tests/seraphis_cache.cpp [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+    for (const auto &checkpoint : boost::adaptors::reverse(m_checkpoints))
+    {
+        window.push_back(checkpoint.first);
```
Comment (UkoeHB) [inline tests/unit_tests/seraphis_cache.cpp]: This seems to contradict what you did in the initialization, if the window is supposed to be sorted in descending order.

File: tests/unit_tests/seraphis_cache.cpp [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+
+        if (window.size() >= m_window_size)
+        {
+            if (m_strategy.should_prune(window, latest_height))
+            {
+                uint64_t middle = window.at(window.size() / 2);
+                m_checkpoints.erase(middle);
+            }
+
+            m_checkpoints.erase(*window.begin());
```
Comment (UkoeHB) [inline tests/unit_tests/seraphis_cache.cpp]: Always erasing the beginning of the window? I don't follow

**Review by UkoeHB — CHANGES_REQUESTED** [review]
I recommend writing a lot of unit tests to work out the bugs and edge cases in this.

File: tests/unit_tests/seraphis_cache.cpp (hunk 3 above)
Comment (UkoeHB) [reply tests/unit_tests/seraphis_cache.cpp]: I'd apply the YAGNI principle here. There is no current or predicted need for multiple pruning strategies.

File: tests/unit_tests/seraphis_cache.cpp (hunk 14 above)
Comment (OfekShochat) [reply tests/unit_tests/seraphis_cache.cpp:62]: Advancing the window
I'll fix the other comments tomorrow. Seems my c++ is incredibly rusty.

File: tests/unit_tests/seraphis_cache.cpp (hunk 4 above)
Comment (ghostway0) [reply tests/unit_tests/seraphis_cache.cpp]: the size when should_prune is called is constant. the other stuff is my stupidity and lack of time spent on this. currently focused on making this work.
for some reason I had the impression that the window is ordered forwards, no idea why. for the other thing, yea, thats why that weird resize (I added that yesterday without any thought put into it.)
will have more brain power over this tomorrow

File: tests/unit_tests/seraphis_cache.cpp (hunk 14 above)
Comment (ghostway0) [reply tests/unit_tests/seraphis_cache.cpp]: advancing the window. this isnt the way to do it, anyway

File: tests/unit_tests/seraphis_cache.cpp (hunk 13 above)
Comment (ghostway0) [reply tests/unit_tests/seraphis_cache.cpp]: yep

File: tests/unit_tests/seraphis_cache.cpp (hunk 4 above)
Comment (UkoeHB) [reply tests/unit_tests/seraphis_cache.cpp]: Assuming an invariant like 'size is nonzero' instead of asserting it makes your code more brittle, and invites nonlocal reasoning.

File: tests/unit_tests/seraphis_cache.cpp (hunk 12 above)
Comment (ghostway0) [reply tests/unit_tests/seraphis_cache.cpp]: I didnt understand what you meant apparently and just put this off as "I don't know what this is".

File: tests/unit_tests/seraphis_cache.cpp (hunk 4 above)
Comment (ghostway0) [reply tests/unit_tests/seraphis_cache.cpp]: ah, I just forgot I was (supposed to be) going from the top; so some of the code is assuming Im going from the bottom and some from the top. sorry for all these, not opening my journey in monero great.

**Review by UkoeHB — APPROVED** [review]
Thanks, I'll take it from here!
